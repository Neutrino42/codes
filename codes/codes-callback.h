/*
 * Copyright (C) 2015 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
*/

#ifndef CODES_CALLBACK_H
#define CODES_CALLBACK_H

#ifdef __cplusplus
extern "C" {
#endif

#include <assert.h>
#include "codes/lp-msg.h"

/* This header defines the following conventions for callback-based event
 * processing (i.e., RPC-like event flow where the callee sends a "return
 * event" to the caller)
 *
 * callers:
 * - give the caller event size (for sanity checking)
 * - use the msg_header convention (lp-msg.h)
 * - specify an integer tag to identify which call the callee return event
 *   corresponds to, in the case of multiple pending callbacks
 * - specify an offset to the return structure, which is defined by the callees
 *
 * callees:
 * - use the resulting set of offsets to write and issue the callback
 *
 * The recommended calling convention for these types of event flows is:
 * void event_func(<params...>,
 *         msg_header const * h, int tag, struct codes_cb_info const * cb);
 */

struct codes_cb_info {
    int event_size;
    int header_offset;
    int tag_offset;
    int cb_ret_offset;
};

/* helper struct for callees - wrap up the typical inputs */
struct codes_cb_params {
    struct codes_cb_info info;
    msg_header h;
    int tag;
};

/* function-like sugar for initializing a codes_cb_info */
#define INIT_CODES_CB_INFO(_cb_info_ptr, _event_type, _header_field, _tag_field, _cb_ret_field) \
    do { \
        (_cb_info_ptr)->event_size    = sizeof(_event_type); \
        (_cb_info_ptr)->header_offset = offsetof(_event_type, _header_field); \
        (_cb_info_ptr)->tag_offset    = offsetof(_event_type, _tag_field); \
        (_cb_info_ptr)->cb_ret_offset = offsetof(_event_type, _cb_ret_field); \
    } while (0)

#define CB_HO(_cb_params_ptr) ((_cb_params_ptr)->info.header_offset)
#define CB_TO(_cb_params_ptr) ((_cb_params_ptr)->info.tag_offset)
#define CB_RO(_cb_params_ptr) ((_cb_params_ptr)->info.cb_ret_offset)

/* Declare return variables at the right byte offsets from a codes_cb_params.
 * Additionally, set the header and tag vars */
#define GET_INIT_CB_PTRS(_cb_params_ptr, _data_ptr, _sender_gid_val, _header_nm, _tag_nm, _rtn_name, _rtn_type) \
    msg_header * _header_nm = \
            (msg_header*)((char*)(_data_ptr) + CB_HO(_cb_params_ptr)); \
    int * _tag_nm = \
            (int*)((char*)(_data_ptr) + CB_TO(_cb_params_ptr));\
    _rtn_type * _rtn_name = \
            (_rtn_type*)((char*)(_data_ptr) + CB_RO(_cb_params_ptr)); \
    msg_set_header((_cb_params_ptr)->h.magic, (_cb_params_ptr)->h.event_type, \
            _sender_gid_val, _header_nm); \
    *_tag_nm = (_cb_params_ptr)->tag

#define SANITY_CHECK_CB(_cb_info_ptr, _ret_type) \
    do { \
        assert((_cb_info_ptr)->event_size > 0 && \
                (_cb_info_ptr)->header_offset >= 0 && \
                (_cb_info_ptr)->tag_offset >= 0 && \
                (_cb_info_ptr)->cb_ret_offset >= 0); \
        size_t _total_size = sizeof(_ret_type) + sizeof(int) + sizeof(msg_header);\
        size_t _esize = (_cb_info_ptr)->event_size; \
        assert(_esize >= _total_size); \
        assert(_esize >= (_cb_info_ptr)->header_offset + sizeof(msg_header)); \
        assert(_esize >= (_cb_info_ptr)->tag_offset + sizeof(int)); \
        assert(_esize >= (_cb_info_ptr)->cb_ret_offset + sizeof(_ret_type)); \
    } while (0)

#ifdef __cplusplus
}
#endif

#endif /* end of include guard: CODES_CALLBACK_H */

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 *  indent-tabs-mode: nil
 * End:
 *
 * vim: ft=c ts=8 sts=4 sw=4 expandtab
 */
