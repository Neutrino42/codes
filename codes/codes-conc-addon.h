/*
 * Copyright (C) 2017 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

#ifndef CODES_CONC_ADDON_H
#define CODES_CONC_ADDON_H

#ifdef __cplusplus
extern "C" {
#endif

#ifdef USE_CONC
#include <ncptl/ncptl.h> 
#endif

#define MAX_CONC_ARGC 20

typedef struct conc_bench_param conc_bench_param;

/* implementation structure */
struct codes_conceptual_bench {
    char *program_name; /* name of the conceptual program */
    int (*conceptual_main)(int* argc, char *argv[]);
};

struct conc_bench_param {
    char *conc_program;
    int conc_argc;
    char *conc_argv[MAX_CONC_ARGC];
};


int codes_conc_bench_load(
        const char* program,
        int* argc, 
        const char *argv[]);


void codes_conceptual_add_bench(struct codes_conceptual_bench const * method);


#ifdef __cplusplus
}
#endif

#endif /* CODES_CONC_ADDON_H */
