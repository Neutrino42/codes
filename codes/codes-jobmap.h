/*
 * Copyright (C) 2015 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

/* SUMMARY - namespace support for multiple groups of ids (jobs) with respect to a
 * flat namespace. Note that this API is meant for static job creation - a more
 * sophisticated method would need to be used to modify job mappings at
 * runtime.
 *
 * Example:
 *
 * job  0      1    2
 * rank 0 1 2  0 1  0 1 2 3
 * ID   0 1 2  3 4  5 6 7 8  (<-- LP relative ID)
 * LP   A B C  D E  F G H I  (<-- provided by codes-mapping)
 * */

#ifndef CODES_JOBMAP_H
#define CODES_JOBMAP_H

/** type markers and parameter defs for jobmaps **/

enum codes_jobmap_type {
    /* the "dummy" jobmap is an example implementation. It simply specifies N
     * jobs, with exactly one rank per job, with a trivial mapping */
    CODES_JOBMAP_DUMMY
};

struct codes_jobmap_params_dummy {
    int num_jobs;
};

/** jobmap interface **/

struct codes_jobmap_ctx;

struct codes_jobmap_id {
    int job;
    int rank; // relative to job
};

struct codes_jobmap_ctx *
codes_jobmap_configure(enum codes_jobmap_type t, void const * params);

void codes_jobmap_destroy(struct codes_jobmap_ctx *c);

struct codes_jobmap_id codes_jobmap_lookup(
        int id,
        struct codes_jobmap_ctx const * c);

int codes_jobmap_get_num_jobs(struct codes_jobmap_ctx const * c);

#endif

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 *  indent-tabs-mode: nil
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
