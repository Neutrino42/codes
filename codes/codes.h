/*
 * Copyright (C) 2013 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */


#ifndef CODES_H
#define CODES_H

#include <ross.h>
#include <assert.h>

#if 0
#define codes_event_new tw_event_new
#else
static inline tw_event * codes_event_new(
    tw_lpid dest_gid, 
    tw_stime offset_ts, 
    tw_lp * sender)
{
    tw_stime abs_ts = offset_ts + tw_now(sender); 
    assert(abs_ts < g_tw_ts_end);
    //printf("codes_event_new() abs_ts: %.9f\n", abs_ts);
    return(tw_event_new(dest_gid, offset_ts, sender));
}
#endif

/* TODO: validate what value we should use here */
/* Modeled latency for communication between local software components and
 * communication between daemons and hardware devices.  Measured in
 * nanoseconds.
 * Modified Jul 7: We want to make sure that the event time stamp generated
is always greater than the default g_tw_lookahead value. Multiplying by 1.1
ensures that if tw_rand_exponential generates a zero time-stamped event, we
still have a timestamp that is greater than g_tw_lookahead. 
 */
#define CODES_MEAN_LOCAL_LATENCY 0.01
static inline tw_stime codes_local_latency(tw_lp *lp)
{
    tw_stime tmp;

    tmp = (1.1 * g_tw_lookahead) + tw_rand_exponential(lp->rng, CODES_MEAN_LOCAL_LATENCY);

    return(tmp);
}

static inline void codes_local_latency_reverse(tw_lp *lp)
{
    tw_rand_reverse_unif(lp->rng);
    return;
}

#endif /* CODES_H */

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ft=c ts=8 sts=4 sw=4 expandtab
 */
