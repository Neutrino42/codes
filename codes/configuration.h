#ifndef __CONFIGURATION_H__
#define __CONFIGURATION_H__

#include <inttypes.h>
#include <mpi.h>
#include "codes/txt_configfile.h"

#define CONFIGURATION_MAX_NAME 256
#define CONFIGURATION_MAX_GROUPS 10
#define CONFIGURATION_MAX_TYPES 10

typedef struct config_lptype_s
{
    char     name[CONFIGURATION_MAX_NAME];
    uint64_t count;
} config_lptype_t;

typedef struct config_lpgroup_s
{
    char     name[CONFIGURATION_MAX_NAME];
    uint64_t repetitions;
    config_lptype_t lptypes[CONFIGURATION_MAX_TYPES];
    uint64_t lptypes_count;
} config_lpgroup_t;

typedef struct config_lpgroups_s
{
    config_lpgroup_t lpgroups[CONFIGURATION_MAX_GROUPS];
    uint64_t lpgroups_count;
} config_lpgroups_t;

/*
 * Load a configuration on the system (collectively)
 *
 * filepath - path and name of file
 * comm     - communicator to distribute file on
 * handle   - output of configuration
 *
 * return 0 on success
 */
int configuration_load (const char * filepath,
                        MPI_Comm comm,
                        ConfigHandle *handle);

/*
 * Free any resources allocated on load.
 *
 * handle - configuration handle
 */
int configuration_free (ConfigHandle *handle);

/*
 * Get's the value for a give section/key pair.
 * Caller's responsible for transforming it to a useable type.
 * Assumes the key name is a KEY configuration type.
 * 
 * handle - configuration handle
 * section_name - name of the section the key is in
 * key_name - name of the key
 * value - pointer to string
 * length - maximum length of string
 */
int configuration_get_value(ConfigHandle *handle,
                            const char * section_name,
                            const char * key_name,
                            char *value,
                            size_t length);

/*
 * Get's the value for a give section/key pair and converts it to an int.
 * Assumes the key name is a KEY configuration type.
 * 
 * handle - configuration handle
 * section_name - name of the section the key is in
 * key_name - name of the key
 * value - returned as a pointer to an integer
 */
int configuration_get_value_int (ConfigHandle *handle,
                                 const char *section_name,
                                 const char *key_name,
                                 int *value);

/*
 * Get's the value for a give section/key pair and converts it to an
 * unsigned int.
 * Assumes the key name is a KEY configuration type.
 * 
 * handle - configuration handle
 * section_name - name of the section the key is in
 * key_name - name of the key
 * value - returned as a pointer to an unsigned integer
 */
int configuration_get_value_uint (ConfigHandle *handle,
                                  const char *section_name,
                                  const char *key_name,
                                  unsigned int *value);

/*
 * Get's the value for a give section/key pair and converts it to a long int.
 * Assumes the key name is a KEY configuration type.
 * 
 * handle - configuration handle
 * section_name - name of the section the key is in
 * key_name - name of the key
 * value - returned as a pointer to a long integer
 */
int configuration_get_value_longint (ConfigHandle *handle,
                                     const char *section_name,
                                     const char *key_name,
                                     long int *value);

/*
 * Get's the value for a give section/key pair and converts it to a double.
 * Assumes the key name is a KEY configuration type.
 * 
 * handle - configuration handle
 * section_name - name of the section the key is in
 * key_name - name of the key
 * value - returned as a pointer to a double
 */
int configuration_get_value_double (ConfigHandle *handle,
                                    const char *section_name,
                                    const char *key_name,
                                    double *value);

/*
 * Get the LPGROUPS configuration from the config file which is stored
 * in the associated data structures.
 *
 * handle - configuration handle
 * section_name - name of section which has the lptypes
 * lpgroups - data structure to hold the lpgroup info
 */
int configuration_get_lpgroups (ConfigHandle *handle,
                                const char *section_name,
                                config_lpgroups_t *lpgroups);

/*
 * Forward reference to the configuration handle
 */
extern ConfigHandle config;

#endif
