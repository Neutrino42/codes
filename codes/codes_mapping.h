/*
 * Copyright (C) 2011, University of Chicago
 *
 * See COPYRIGHT notice in top-level directory.
 */

/* SUMMARY:
 * CODES custom mapping file for ROSS
 */
#include "configuration.h"
#include "codes.h"
#include "lp-type-lookup.h"
#define MAX_NAME_LENGTH 256

/* number of LPs assigned to the current PE (abstraction of MPI rank) */
static int lps_for_this_pe = 0;

/* data structure to hold configuration */
configuration_t config;

/* char arrays for holding lp type name and group name*/
char local_grp_name[MAX_NAME_LENGTH], local_lp_name[MAX_NAME_LENGTH];

/* Returns number of LPs on the current PE */
int codes_mapping_get_lps_for_pe(void);

/* Takes the global LP ID and returns the rank (PE id) on which the LP is mapped.*/
tw_peid codes_mapping( tw_lpid gid);

/* loads the configuration file and sets up the number of LPs on each PE. */
void codes_mapping_setup(char* filepath);

/* Takes the group name , type name, rep ID and offset (for that lp type + repetition) and then returns the global LP ID. */
void codes_mapping_get_lp_id(char* grp_name, char* lp_type_name, int rep_id, int offset, tw_lpid* gid);

/* takes the LP ID and returns its grp name and index, lp type name and ID, repetition ID and the offset of the LP 
 * (for multiple LPs in a repetition). */
void codes_mapping_get_lp_info(tw_lpid gid, char* grp_name, int* grp_id, int* lp_type_id, char* lp_type_name, int* grp_rep_id, int* offset);

/* assigns local and global lp ids for ROSS. */
void codes_mapping_init(void);

/* Takes the global LP ID, maps it to the local LP ID and returns the LP.
 * lps have global and local LP IDs. 
 * global LP IDs are unique across all PEs, local LP IDs are unique within a PE. */
tw_lp * codes_mapping_to_lp( tw_lpid lpid);

