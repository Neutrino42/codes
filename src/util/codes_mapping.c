/*
 * Copyright (C) 2013 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

/* SUMMARY:
 * CODES custom mapping file for ROSS
 */
#include "codes/codes_mapping.h"

#define CODES_MAPPING_DEBUG 0

/* number of LPs assigned to the current PE (abstraction of MPI rank).
 * for lp counts which are not divisible by the number of ranks, keep 
 * modulus around */
static int lps_per_pe_floor = 0;
static int lps_leftover = 0;

static int mem_factor = 256;

static inline int mini(int a, int b){ return a < b ? a : b; }

/* char arrays for holding lp type name and group name*/
char local_grp_name[MAX_NAME_LENGTH], local_lp_name[MAX_NAME_LENGTH];

config_lpgroups_t lpconf;

int codes_mapping_get_lps_for_pe()
{
    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
#if CODES_MAPPING_DEBUG
    printf("%d lps for rank %d\n", lps_per_pe_floor+(g_tw_mynode < lps_leftover), rank);
#endif
  return lps_per_pe_floor + (g_tw_mynode < lps_leftover);
}

/* Takes the global LP ID and returns the rank (PE id) on which the LP is mapped */
tw_peid codes_mapping( tw_lpid gid)
{
    int lps_on_pes_with_leftover = lps_leftover * (lps_per_pe_floor+1);
    if (gid < lps_on_pes_with_leftover){
        return gid / (lps_per_pe_floor+1);
    }
    else{
        return (gid-lps_on_pes_with_leftover)/lps_per_pe_floor + lps_leftover;
    }
  /*return gid / lps_per_pe_floor;*/
}

int codes_mapping_get_group_reps(char* grp_name)
{
  int grp;
  for(grp = 0; grp < lpconf.lpgroups_count; grp++)
  {
     if(strcmp(lpconf.lpgroups[grp].name, grp_name) == 0)
	     return lpconf.lpgroups[grp].repetitions;
  }
  return -1;
}
int codes_mapping_get_lp_count(char* grp_name, char* lp_type_name)
{
   int grp, lpt, lp_types_count;

// Account for all lps in the previous groups 
  for(grp = 0; grp < lpconf.lpgroups_count; grp++)
   {
       lp_types_count = lpconf.lpgroups[grp].lptypes_count;
        
       if(strcmp(lpconf.lpgroups[grp].name, grp_name) == 0)
	{
	 for(lpt = 0; lpt < lp_types_count; lpt++)
 	 {
	    if(strcmp(lpconf.lpgroups[grp].lptypes[lpt].name, lp_type_name) == 0)
		return lpconf.lpgroups[grp].lptypes[lpt].count;
	 }		 
	}
   }
  return -1;
}

int codes_mapping_get_global_lp_count(char* lp_type_name){
    int total_ct = 0;
    int grp;
    for (grp = 0; grp < lpconf.lpgroups_count; grp++){
        int lpt;
        int lp_types_count = lpconf.lpgroups[grp].lptypes_count;
        int reps = lpconf.lpgroups[grp].repetitions;
        config_lptype_t *c = lpconf.lpgroups[grp].lptypes;
        for (lpt = 0; lpt < lp_types_count; lpt++){
            if(strcmp(c[lpt].name, lp_type_name) == 0){
                total_ct += c[lpt].count * reps;
            }
        }
    }
    return total_ct;
}

/* This function takes the group ID , type ID and rep ID then returns the global LP ID */
/* TODO: Add string based search for LP group and type names */
void codes_mapping_get_lp_id(char* grp_name, char* lp_type_name, int rep_id, int offset, tw_lpid* gid)
{
 int grp, lpt, lpcount = 0, lp_types_count, rep, count_for_this_lpt;
 short found = 0;

 // Account for all lps in the previous groups 
 for(grp = 0; grp < lpconf.lpgroups_count; grp++)
  {
    lp_types_count = lpconf.lpgroups[grp].lptypes_count;
    rep = lpconf.lpgroups[grp].repetitions;

    if(strcmp(lpconf.lpgroups[grp].name, grp_name) == 0)
    {
	    found = 1;
	    break;
    }
    
    for(lpt = 0; lpt < lp_types_count; lpt++)
      lpcount += (rep * lpconf.lpgroups[grp].lptypes[lpt].count);
  }

 assert(found);
 found = 0;

 lp_types_count = lpconf.lpgroups[grp].lptypes_count;
 
 // Account for the previous lp types in the current repetition
 for(lpt = 0; lpt < lp_types_count; lpt++)
 {
   count_for_this_lpt = lpconf.lpgroups[grp].lptypes[lpt].count;

   if(strcmp(lpconf.lpgroups[grp].lptypes[lpt].name, lp_type_name) == 0)
   {
     found = 1;
     break;
   }
   
   lpcount += count_for_this_lpt;
 }

 assert(found);
 // Account for all previous repetitions
 for(rep = 0; rep < rep_id; rep++)
 {
    for(lpt = 0; lpt < lp_types_count; lpt++)
      lpcount += lpconf.lpgroups[grp].lptypes[lpt].count;
 }
   *gid = lpcount + offset % count_for_this_lpt;
}

int codes_mapping_get_lp_global_rel_id(tw_lpid gid){
    /* use mapping function to get our starting point */
    char grp_name[MAX_NAME_LENGTH];
    char lp_type_name[MAX_NAME_LENGTH];
    int grp_id, lp_type_id, grp_rep_id, offset;
    codes_mapping_get_lp_info(gid, grp_name, &grp_id, &lp_type_id, lp_type_name,
            &grp_rep_id, &offset);

    /* now, go through the groups that occurred before and count up the lps of
     * the same name */
    int grp;
    int total_lp_count = 0;
    for (grp = 0; grp < grp_id; grp++){
        int rep = lpconf.lpgroups[grp].repetitions;
        int l;
        /* get the total number of LPs in this group */
        for (l = 0; l < lpconf.lpgroups[grp].lptypes_count; l++){
            if (strcmp(lp_type_name, lpconf.lpgroups[grp].lptypes[l].name) == 0){
                total_lp_count += rep * lpconf.lpgroups[grp].lptypes[l].count;
            }
        }
    }

    /* the global relative lp id is my relative position within the group +
     * the number of lps that came before */
    int nlp_per_rep = lpconf.lpgroups[grp_id].lptypes[lp_type_id].count;
    return total_lp_count + (grp_rep_id * nlp_per_rep) + offset;
}


/* This function takes the LP ID and returns its grp index, lp type ID and repetition ID */
void codes_mapping_get_lp_info(tw_lpid gid, char* grp_name, int* grp_id, int* lp_type_id, char* lp_type_name, int* grp_rep_id, int* offset)
{
  int grp, lpt, rep, grp_offset, lp_offset, rep_offset;
  int lp_tmp_id, lp_types_count, lp_count;
  unsigned long grp_lp_count=0;
  short found = 0;
 
  /* Find the group id first */ 
  for(grp = 0; grp < lpconf.lpgroups_count; grp++)
  {
    grp_offset = 0;
    rep_offset = 0;
    rep = lpconf.lpgroups[grp].repetitions;
    lp_types_count = lpconf.lpgroups[grp].lptypes_count;

    for(lpt = 0; lpt < lp_types_count; lpt++)
    {
	lp_count = lpconf.lpgroups[grp].lptypes[lpt].count;
	grp_offset += (rep * lp_count);
	rep_offset += lp_count;
    }
    /* Each gid is assigned an increasing number starting from 0th group and 0th lp type
     * so we check here if the gid lies within the numeric range of a group */ 
    if(gid >= grp_lp_count && gid < grp_lp_count + grp_offset)
    {
	*grp_id = grp;
	 strcpy(local_grp_name, lpconf.lpgroups[grp].name);
	 lp_offset = gid - grp_lp_count; /* gets the lp offset starting from the point where the group begins */
         *grp_rep_id = lp_offset / rep_offset;
          lp_tmp_id = lp_offset - (*grp_rep_id * rep_offset);
	  found = 1;
	 break;
    }
    grp_lp_count += grp_offset; /* keep on increasing the group lp count to next group range*/
  }
  assert(found);

  lp_offset = 0;
  found = 0; /* reset found for finding LP type */
 /* Now we compute the LP type ID here based on the lp offset that we just got */ 
  for(lpt = 0; lpt < lp_types_count; lpt++)
  {
     lp_count = lpconf.lpgroups[grp].lptypes[lpt].count;
     if(lp_tmp_id >= lp_offset && lp_tmp_id < lp_offset + lp_count)
     {
	     *lp_type_id = lpt;
	      strcpy(local_lp_name, lpconf.lpgroups[grp].lptypes[lpt].name);
	      *offset = lp_tmp_id - lp_offset;
	      found = 1;
     	      break;
     }
     lp_offset += lp_count;
  }
  assert(found);
  strncpy(grp_name, local_grp_name, MAX_NAME_LENGTH);
  strncpy(lp_type_name, local_lp_name, MAX_NAME_LENGTH);
  //printf("\n gid %d lp type name %s rep_id %d ", gid, lp_type_name, *grp_rep_id);
}

/* This function assigns local and global LP Ids to LPs */
static void codes_mapping_init(void)
{
     int grp_id, lpt_id, rep_id, offset;
     tw_lpid ross_gid, ross_lid; /* ross global and local IDs */
     tw_pe * pe;
     char lp_type_name[MAX_NAME_LENGTH];
     char grp_name[MAX_NAME_LENGTH];
     int nkp_per_pe = g_tw_nkp;
     tw_lpid         lpid, kpid;

     /* have 16 kps per pe, this is the optimized configuration for ROSS custom mapping */
     for(kpid = 0; kpid < nkp_per_pe; kpid++)
	tw_kp_onpe(kpid, g_tw_pe[0]);

     int lp_start =
         g_tw_mynode * lps_per_pe_floor + mini(g_tw_mynode,lps_leftover);
     int lp_end =
         (g_tw_mynode+1) * lps_per_pe_floor + mini(g_tw_mynode+1,lps_leftover);
     codes_mapping_get_lp_info(lp_start, grp_name, &grp_id, &lpt_id, lp_type_name, &rep_id, &offset);

     for (lpid = lp_start; lpid < lp_end; lpid++)
      {
	 ross_gid = lpid;
	 ross_lid = lpid - lp_start;
	 kpid = ross_lid % g_tw_nkp;
	 pe = tw_getpe(kpid % g_tw_npe);
	 codes_mapping_get_lp_info(ross_gid, grp_name, &grp_id, &lpt_id, lp_type_name, &rep_id, &offset);
#if CODES_MAPPING_DEBUG
         printf("lp:%lu --> kp:%lu, pe:%llu\n", ross_gid, kpid, pe->id);
#endif
	 tw_lp_onpe(ross_lid, pe, ross_gid);
	 tw_lp_onkp(g_tw_lp[ross_lid], g_tw_kp[kpid]);
	 tw_lp_settype(ross_lid, lp_type_lookup(lp_type_name));
     }
     return;
}

/* This function takes the global LP ID, maps it to the local LP ID and returns the LP 
 * lps have global and local LP IDs
 * global LP IDs are unique across all PEs, local LP IDs are unique within a PE */
static tw_lp * codes_mapping_to_lp( tw_lpid lpid)
{
   int index = lpid - (g_tw_mynode * lps_per_pe_floor) -
       mini(g_tw_mynode, lps_leftover);
//   printf("\n global id %d index %d lps_before %d lps_offset %d local index %d ", lpid, index, lps_before, g_tw_mynode, local_index);
   return g_tw_lp[index];
}

/* This function loads the configuration file and sets up the number of LPs on each PE */
void codes_mapping_setup()
{
  int grp, lpt, message_size;
  int pes = tw_nnodes();

  configuration_get_lpgroups(&config, "LPGROUPS", &lpconf);

  for (grp = 0; grp < lpconf.lpgroups_count; grp++)
   {
    for (lpt = 0; lpt < lpconf.lpgroups[grp].lptypes_count; lpt++)
	lps_per_pe_floor += (lpconf.lpgroups[grp].lptypes[lpt].count * lpconf.lpgroups[grp].repetitions);
   }
  lps_leftover = lps_per_pe_floor % pes;
  lps_per_pe_floor /= pes;
 //printf("\n LPs for this PE are %d reps %d ", lps_per_pe_floor,  lpconf.lpgroups[grp].repetitions);
  g_tw_mapping=CUSTOM;
  g_tw_custom_initial_mapping=&codes_mapping_init;
  g_tw_custom_lp_global_to_local_map=&codes_mapping_to_lp;
  g_tw_events_per_pe = mem_factor * codes_mapping_get_lps_for_pe();
  configuration_get_value_int(&config, "PARAMS", "message_size", &message_size);
  if(!message_size)
  {
      message_size = 256;
      printf("\n Warning: ross message size not defined, resetting it to %d", message_size);
  }
  tw_define_lps(codes_mapping_get_lps_for_pe(), message_size, 0);
}


/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
