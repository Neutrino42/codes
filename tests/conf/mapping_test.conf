LPGROUPS 
{
    GRP1
    {
        repetitions="2";
        a="1";
        b="2";
    }
    GRP2
    {
        repetitions="3";
        c="1";
        b="1";
    }
}

PARAMS
{
    message_size="256";
}
