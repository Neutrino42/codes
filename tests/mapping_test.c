/*
 * Copyright (C) 2014 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

#include <ross.h>
#include "codes/configuration.h"
#include "codes/codes_mapping.h"
#include "codes/codes.h"

/* define three types of lps for mapping test */

typedef struct a_state_s{
    int id;
} a_state;
typedef a_state b_state;
typedef a_state c_state;

static void a_init(a_state *ns, tw_lp *lp){
    ns->id = codes_mapping_get_lp_global_rel_id(lp->gid);
}
static void b_init(b_state *ns, tw_lp *lp){ a_init(ns,lp); }
static void c_init(c_state *ns, tw_lp *lp){ a_init(ns,lp); }

static void a_finalize(a_state *ns, tw_lp *lp){
    printf("TEST2 %lu %d a\n", lp->gid, ns->id);
}
static void b_finalize(b_state *ns, tw_lp *lp){
    printf("TEST2 %lu %d b\n", lp->gid, ns->id);
}
static void c_finalize(c_state *ns, tw_lp *lp){
    printf("TEST2 %lu %d c\n", lp->gid, ns->id);
}

tw_lptype a_lp = {
     (init_f) a_init,
     (event_f) NULL,
     (revent_f) NULL,
     (final_f)  a_finalize, 
     (map_f) codes_mapping,
     sizeof(a_state),
};
tw_lptype b_lp = {
     (init_f) b_init,
     (event_f) NULL,
     (revent_f) NULL,
     (final_f)  b_finalize, 
     (map_f) codes_mapping,
     sizeof(b_state),
};
tw_lptype c_lp = {
     (init_f) c_init,
     (event_f) NULL,
     (revent_f) NULL,
     (final_f)  c_finalize, 
     (map_f) codes_mapping,
     sizeof(c_state),
};

static char conf_file_name[128] = {'\0'};
static const tw_optdef app_opt [] =
{
	TWOPT_GROUP("codes-mapping test case" ),
    TWOPT_CHAR("codes-config", conf_file_name, "name of codes configuration file"),
	TWOPT_END()
};

int main(int argc, char *argv[])
{
    tw_opt_add(app_opt);
    tw_init(&argc, &argv);

    if (!conf_file_name[0]){
        fprintf(stderr, "Expected \"codes-config\" option, please see --help.\n");
        MPI_Finalize();
        return 1;
    }
    if (configuration_load(conf_file_name, MPI_COMM_WORLD, &config)){
        fprintf(stderr, "Error loading config file %s.\n", conf_file_name);
        MPI_Finalize();
        return 1;
    }

    lp_type_register("a", &a_lp);
    lp_type_register("b", &b_lp);
    lp_type_register("c", &c_lp);

    codes_mapping_setup();

    printf("TEST1 %d a\n", codes_mapping_get_global_lp_count("a"));
    printf("TEST1 %d b\n", codes_mapping_get_global_lp_count("b"));
    printf("TEST1 %d c\n", codes_mapping_get_global_lp_count("c"));

    tw_run();
    tw_end();

    return 0;
}

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ft=c ts=8 sts=4 sw=4 expandtab
 */
