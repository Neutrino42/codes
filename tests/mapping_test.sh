#!/bin/bash

tst=$srcdir/tests
set -e
tests/mapping_test --sync=1 --codes-config=$tst/conf/mapping_test.conf | \
    grep TEST | sort -s -k 1d,1d -k 2n,2n > mapping_test.out

diff $tst/expected/mapping_test.out mapping_test.out
err=$?

if [ "$err" -eq 0 ]; then
    rm mapping_test.out
fi
