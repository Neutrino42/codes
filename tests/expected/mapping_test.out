TEST1 2 a
TEST1 3 c
TEST1 7 b
TEST2 0 0 a
TEST2 1 0 b
TEST2 2 1 b
TEST2 3 1 a
TEST2 4 2 b
TEST2 5 3 b
TEST2 6 0 c
TEST2 7 4 b
TEST2 8 1 c
TEST2 9 5 b
TEST2 10 2 c
TEST2 11 6 b
